File: Plug-Ins/Template Editor/TemplateWindowController.m
#import "TemplateWindowController.h"
#import "ResourceStream.h"
#import "GroupElement.h"
#import "ElementLSTB.h"

@implementation TemplateWindowController
@synthesize dataList;
@synthesize resource;
@synthesize tmpl;

- (instancetype)initWithResource:(Resource *)newResource
{
	return nil;
}

- (instancetype)initWithResource:(Resource *)newResource template:(Resource *)tmplResource
{
	self = [self initWithWindowNibName:@"TemplateWindow"];
	if (!self) {
		return nil;
	}
	
	//undoManager = [[NSUndoManager alloc] init];
    resource = newResource;
    tmpl = tmplResource;
    resourceStructure = nil;
    [self loadResource];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(resourceDataDidChange:) name:@"ResourceDataDidChangeNotification" object:resource];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(templateDataDidChange:) name:@"ResourceDataDidChangeNotification" object:tmpl];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(windowDidBecomeKey:) name:NSWindowDidBecomeKeyNotification object:self.window];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(windowDidResignKey:) name:NSWindowDidResignKeyNotification object:self.window];
	return self;

}

- (void)windowDidLoad
{
	[super windowDidLoad];
    [self.window setTitle:resource.defaultWindowTitle];
    [dataList expandItem:nil expandChildren:YES];
}

- (void)templateDataDidChange:(NSNotification *)notification
{
    // Reload the template while keeping the current data
    NSData *currentData = [resourceStructure getResourceData];
    [self readTemplate];
    [resourceStructure readResourceData:currentData];
    // expand all
    [dataList reloadData];
    [dataList expandItem:nil expandChildren:YES];
}

- (void)resourceDataDidChange:(NSNotification *)notification
{
    if (!self.window.isDocumentEdited)
        [self loadResource];
}

- (void)readTemplate
{
    NSInputStream *stream = [[NSInputStream alloc] initWithData:tmpl.data];
    resourceStructure = [ElementList listFromStream:stream];
    resourceStructure.controller = self;
    [resourceStructure configureElements];
}

- (void)loadResource
{
    [self readTemplate];
    [resourceStructure readResourceData:resource.data];
	// expand all
    [dataList reloadData];
	[dataList expandItem:nil expandChildren:YES];
    self.documentEdited = NO;
}

- (IBAction)saveResource:(id)sender
{
    // send the new resource data to ResKnife
    [[NSNotificationCenter defaultCenter] removeObserver:self name:@"ResourceDataDidChangeNotification" object:resource];
    resource.data = [resourceStructure getResourceData];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(resourceDataDidChange:) name:@"ResourceDataDidChangeNotification" object:resource];
    self.documentEdited = NO;
}

- (IBAction)revertResource:(id)sender
{
    [self loadResource];
}

#pragma mark -
#pragma mark Table Management

- (NSView *)outlineView:(NSOutlineView *)outlineView viewForTableColumn:(NSTableColumn *)tableColumn item:(__kindof Element *)item
{
    if ([tableColumn.identifier isEqualToString:@"data"]) {
        NSView *view = [[NSView alloc] initWithFrame:NSMakeRect(0, 0, tableColumn.width, item.rowHeight)];
        [item configureView:view];
        view.toolTip = item.tooltip;
        return view;
    } else if (tableColumn) {
        NSString *identifier = tableColumn.identifier;
        if (item.class == ElementLSTB.class && [item allowsCreateListEntry])
            identifier = @"listLabel";
        NSTableCellView *view = [outlineView makeViewWithIdentifier:identifier owner:self];
        view.textField.stringValue = item.displayLabel;
        view.toolTip = item.tooltip;
        return view;
    } else {
        NSTableCellView *view = [outlineView makeViewWithIdentifier:@"groupView" owner:self];
        [item configureGroupView:view];
        return view;
    }
}

- (id)outlineView:(NSOutlineView*)outlineView child:(NSInteger)index ofItem:(Element *)item
{
	if (item == nil)
        return [resourceStructure elementAtIndex:index];
	else
        return [item subElementAtIndex:index];
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isItemExpandable:(Element *)item
{
    return item.hasSubElements;
}

- (NSInteger)outlineView:(NSOutlineView *)outlineView numberOfChildrenOfItem:(Element *)item
{
	if (item == nil)
		return resourceStructure.count;
	else
        return item.subElementCount;
}

- (double)outlineView:(NSOutlineView *)outlineView heightOfRowByItem:(Element *)item
{
    return item.rowHeight;
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isGroupItem:(Element *)item
{
    return [item conformsToProtocol:@protocol(GroupElement)];
}

#pragma mark -
#pragma mark Menu Management

- (IBAction)itemValueUpdated:(id)sender
{
    self.documentEdited = YES;
}

// these next five methods are a crude hack - the items ought to be in the responder chain themselves
- (IBAction)createNewItem:(id)sender;
{
	// This works by selecting a list element (LSTB) and passing the message on to it
    NSInteger row = [dataList rowForView:(NSView *)self.window.firstResponder];
    id element = [dataList itemAtRow:row];
	if ([element class] == ElementLSTB.class && [element allowsCreateListEntry]) {
        [element createListEntry];
		[dataList reloadData];
        NSView *newHeader = [dataList viewAtColumn:0 row:row makeIfNecessary:YES];
        [self.window makeFirstResponder:newHeader];
        self.documentEdited = YES;
        // Expand the item and scroll the new content into view
		[dataList expandItem:[dataList itemAtRow:row] expandChildren:YES];
        NSView *lastChild = [dataList rowViewAtRow:[dataList rowForItem:element] makeIfNecessary:YES];
        [lastChild scrollRectToVisible:lastChild.bounds];
        [newHeader scrollRectToVisible:newHeader.superview.bounds];
	}
}

- (IBAction)delete:(id)sender;
{
    NSInteger row = [dataList rowForView:(NSView *)self.window.firstResponder];
    id element = [dataList itemAtRow:row];
    if ([element class] == ElementLSTB.class && [element allowsRemoveListEntry]) {
        [element removeListEntry];
        [dataList reloadData];
        [self.window makeFirstResponder:[dataList viewAtColumn:0 row:row makeIfNecessary:YES]];
        self.documentEdited = YES;
    }
}

- (IBAction)cut:(id)sender;
{
}

- (IBAction)copy:(id)sender;
{
}

- (IBAction)paste:(id)sender;
{
}

- (BOOL)validateMenuItem:(NSMenuItem *)item
{
    id element = [dataList itemAtRow:[dataList rowForView:(NSView *)self.window.firstResponder]];
	if (item.action == @selector(createNewItem:))
        return [element class] == ElementLSTB.class && [element allowsCreateListEntry];
	else if (item.action == @selector(delete:))
        return [element class] == ElementLSTB.class && [element allowsRemoveListEntry];
	else if (item.action == @selector(saveResource:) || item.action == @selector(revertResource:))
        return self.window.documentEdited;
	else
        return NO;
}

- (NSString *)windowTitleForDocumentDisplayName:(NSString *)displayName
{
    return resource.defaultWindowTitle;
}

- (void)windowDidBecomeKey:(NSNotification *)notification
{
	NSMenu *resourceMenu = [[[NSApp mainMenu] itemWithTag:3] submenu];
	NSMenuItem *createItem = [resourceMenu itemWithTag:0];
	[createItem setTitle:NSLocalizedString(@"Create List Entry", nil)];
}

- (void)windowDidResignKey:(NSNotification *)notification
{
	NSMenu *resourceMenu = [[[NSApp mainMenu] itemWithTag:3] submenu];
	NSMenuItem *createItem = [resourceMenu itemWithTag:0];
	[createItem setTitle:NSLocalizedString(@"Create New Resource…", nil)];
}

@end

#pragma mark -
#pragma mark Table Event Handling

// Table views don't support tabbing between rows so we need to handle the key view loop manually
@implementation NTOutlineView

- (void)selectPreviousKeyView:(id)sender
{
    NSView *view = (NSView *)self.window.firstResponder;
    if (view.previousValidKeyView && view.previousValidKeyView != self) {
        [self.window selectPreviousKeyView:view];
        return;
    }
    NSInteger row = [self rowForView:view];
    // Loop through all rows. Break if we come back to where we started without having found anything.
    for (NSInteger i = row-1; i != row; i--) {
        if (i == -1) continue;
        if (i == -2) {
            i = self.numberOfRows;
            continue;
        }
        // Going backward we need to look at the column view and see if it's valid
        view = [self viewAtColumn:1 row:i makeIfNecessary:YES];
        if (![view canBecomeKeyView]) view = [view.subviews lastObject];
        if (![view canBecomeKeyView]) view = [self viewAtColumn:0 row:i makeIfNecessary:YES];
        if ([view canBecomeKeyView]) {
            [self.window makeFirstResponder:view];
            [view scrollRectToVisible:view.superview.bounds];
            return;
        }
    }
}

- (void)selectNextKeyView:(id)sender
{
    NSView *view = (NSView *)self.window.firstResponder;
    if (view.nextValidKeyView) {
        [self.window selectNextKeyView:view];
        return;
    }
    NSInteger row = [self rowForView:view];
    for (NSInteger i = row+1; i != row; i++) {
        if (i == -1) continue;
        if (i == self.numberOfRows) {
            i = -2;
            continue;
        }
        // Going forward we can ask the row for its nextValidKeyView
        view = [self rowViewAtRow:i makeIfNecessary:YES].nextValidKeyView;
        if (view) {
            [self.window makeFirstResponder:view];
            [view scrollRectToVisible:view.superview.bounds];
            return;
        }
    }
}

// Create a little more space for labels by removing the disclosure triangles
- (NSRect)frameOfOutlineCellAtRow:(NSInteger)row
{
    return NSZeroRect;
}

@end

#pragma mark -
#pragma mark List Header Handling

// This view allows focusing on an LSTB label so we can add/remove list entries
@implementation NTFocusView

- (BOOL)acceptsFirstResponder
{
    return YES;
}

- (void)mouseDown:(NSEvent *)event
{
    [self.window makeFirstResponder:self];
}

- (NSRect)focusRingMaskBounds
{
    return self.textField.frame;
}

- (void)drawFocusRingMask
{
    NSRectFill(self.focusRingMaskBounds);
}

@end

File: Plug-Ins/Template Editor/TemplateWindowController.swift
import Cocoa
import RFSupport

class TemplateWindowController: AbstractEditor, TemplateEditor, NSOutlineViewDataSource, NSOutlineViewDelegate {
    static let supportedTypes: [String] = []
    
    let resource: Resource
    let manager: RFEditorManager
    let template: Resource
    private let filter: TemplateFilter.Type?
    private var resourceStructure: ElementList!
    private var validStructure = false
    @IBOutlet var dataList: TabbableOutlineView!
    var resourceCache: [String: [Resource]] = [:] // Shared cache for RSID/CASR
    
    override var windowNibName: String {
        return "TemplateWindow"
    }
    
    class func parseBasicTemplate(_ template: Resource, manager: RFEditorManager) throws -> [TemplateField] {
        return try TemplateParser(template: template, manager: manager, basic: true).parse()
    }
    
    required init?(resource: Resource, manager: RFEditorManager, template: Resource, filter: TemplateFilter.Type?) {
        self.resource = resource
        self.manager = manager
        self.template = template
        self.filter = filter
        super.init(window: nil)
        
        if !self.load(data: resource.data) {
            return nil
        }
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.resourceDataDidChange(_:)), name: .ResourceDataDidChange, object: resource)
        NotificationCenter.default.addObserver(self, selector: #selector(self.templateDataDidChange(_:)), name: .ResourceDataDidChange, object: template)
    }

    required init(resource: Resource, manager: RFEditorManager) {
        fatalError("init(resource:manager:) has not been implemented")
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    @objc func resourceDataDidChange(_ notification: NSNotification) {
        if self.window?.isDocumentEdited != true {
            _ = self.load(data: resource.data)
        }
    }
    
    @objc func templateDataDidChange(_ notification: NSNotification) {
        // Reload the template while keeping the current data
        _ = self.load(data: resourceStructure.getResourceData())
        dataList.expandItem(nil, expandChildren: true)
    }
        
    override func windowDidLoad() {
        self.window?.title = resource.defaultWindowTitle
        dataList.expandItem(nil, expandChildren: true)
        if validStructure && resource.data.isEmpty {
            self.setDocumentEdited(true)
        }
    }
    
    func load(data: Data) -> Bool {
        resourceStructure = ElementList(controller: self)
        if let filter = filter {
            resourceStructure.insert(ElementDVDR(type: "DVDR", label: "Filter Enabled: \(filter.name)"))
        }
        validStructure = resourceStructure.readTemplate(template)
        if validStructure && !resource.data.isEmpty {
            do {
                let data = try filter?.filter(data: resource.data, for: resource.typeCode) ?? resource.data
                let reader = BinaryDataReader(data)
                try resourceStructure.readData(from: reader)
                if reader.remainingBytes > 0 {
                    // Show warning
                    NSApp.presentError(TemplateError.truncate)
                }
            } catch BinaryDataReaderError.insufficientData {
                // Ignore error, data will be padded on save
            } catch let error {
                NSApp.presentError(error)
                return false
            }
        }
        // expand all
        dataList?.reloadData()
        dataList?.expandItem(nil, expandChildren: true)
        return true
    }
    
    // MARK: - Helper functions
    
    // This function is used by RSID/CASR to get resources for the combo box, caching them by type to improve performance.
    // Note that the cache is on the window controller instance rather than static, to ensure the resources aren't retained indefinitely.
    func resources(ofType type: String) -> [Resource] {
        if resourceCache[type] == nil {
            // Find resources in all documents, deduplicate and sort by name
            let allResources = manager.allResources(ofType: ResourceType(type, resource.typeAttributes), currentDocumentOnly: false)
            let resources: [Int: Resource] = allResources.reduce(into: [:]) { (result, resource) in
                if result[resource.id] == nil && !resource.name.isEmpty {
                    result[resource.id] = resource
                }
            }
            resourceCache[type] = resources.values.sorted {
                $0.name.localizedStandardCompare($1.name) == .orderedAscending
            }
        }
        return resourceCache[type]!
    }
    
    func openOrCreateResource(typeCode: String, id: Int) {
        let type = ResourceType(typeCode, resource.typeAttributes)
        if let resource = manager.findResource(type: type, id: id, currentDocumentOnly: false) {
            manager.open(resource: resource)
        } else {
            manager.createResource(type: type, id: id, name: "")
        }
    }
    
    // MARK: - OutlineView functions
    
    func outlineView(_ outlineView: NSOutlineView, viewFor tableColumn: NSTableColumn?, item: Any) -> NSView? {
        let item = item as! Element
        let view: NSView
        if let tableColumn = tableColumn {
            var identifier = tableColumn.identifier
            if identifier.rawValue == "data" {
                view = DataView(frame: NSMakeRect(0, 2, tableColumn.width, CGFloat(item.rowHeight)))
                item.configure(view: view)
            } else {
                // Use the focusable list label for elements that allow creating entries
                if let item = item as? ElementLSTB, item.allowsCreateListEntry() {
                    identifier = NSUserInterfaceItemIdentifier(item.allowsRemoveListEntry() ? "listLabel" : "listEndLabel")
                }
                view = outlineView.makeView(withIdentifier: identifier, owner: self)!
                let textField = (view as! NSTableCellView).textField!
                textField.stringValue = item.displayLabel
                textField.alignment = item is ElementLSTB ? .left : .right
                textField.allowsDefaultTighteningForTruncation = true
            }
            view.toolTip = item.tooltip
        } else {
            view = outlineView.makeView(withIdentifier: NSUserInterfaceItemIdentifier("groupView"), owner: self)!
            (item as! GroupElement).configureGroup(view: view as! NSTableCellView)
            if item is CounterElement {
                // Match indentation of list headers with that in TabbableOutlineView
                let indent = outlineView.level(forItem: item) * 16
                let constraint = view.constraints.first(where: { $0.firstAttribute == .leading })!
                constraint.constant = 5 + CGFloat(indent)
            }
        }
        return view
    }
    
    func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
        if let item = item as? CollectionElement {
            return item.subElementCount
        }
        return Int(resourceStructure.count)
    }
    
    func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
        if let item = item as? CollectionElement {
            return item.subElement(at: index)
        }
        return resourceStructure.element(at: index)
    }
    
    func outlineView(_ outlineView: NSOutlineView, isItemExpandable item: Any) -> Bool {
        return item is CollectionElement
    }
    
    func outlineView(_ outlineView: NSOutlineView, heightOfRowByItem item: Any) -> CGFloat {
        return CGFloat((item as! Element).rowHeight) + 4
    }
    
    func outlineView(_ outlineView: NSOutlineView, isGroupItem item: Any) -> Bool {
        return item is GroupElement
    }
    
    // MARK: - Menu functions
    
    @IBAction func saveResource(_ sender: Any) {
        if self.window?.makeFirstResponder(dataList) != false {
            do {
                let data = resourceStructure.getResourceData()
                resource.data = try filter?.unfilter(data: data, for: resource.typeCode) ?? data
                self.setDocumentEdited(false)
            } catch let error {
                NSApp.presentError(error)
            }
        }
    }
    
    @IBAction func revertResource(_ sender: Any) {
        _ = self.load(data: resource.data)
        self.setDocumentEdited(false)
    }
    
    @IBAction func itemValueUpdated(_ sender: Any) {
        self.setDocumentEdited(true)
    }
    
    func windowDidBecomeKey(_ notification: Notification) {
        let createItem = NSApp.mainMenu?.item(withTag: 3)?.submenu?.item(withTag: 0)
        createItem?.title = NSLocalizedString("Create List Entry", comment: "")
    }
    
    func windowDidResignKey(_ notification: Notification) {
        let createItem = NSApp.mainMenu?.item(withTag: 3)?.submenu?.item(withTag: 0)
        createItem?.title = NSLocalizedString("Create New Resource…", comment: "")
    }
}
